docs/adr/0001-add-generic-fetcher.md

```diff
@@ -0,0 +1,109 @@
+# Add generic fetcher
+
+- Status: proposed
+- Date: 2024-10-30
+
+## Context
+
+The main motivation for this change is to cover use cases of users that need to download arbitrary files that don't fit
+within an established package ecosystem cachi2 could potentially otherwise support. The target audience is users that
+want to use cachi2 to achieve hermetic builds and want an easy way to also include these arbitrary files, that cachi2
+will account for in the SBOM it produces.
+
+## Decision
+
+This change introduces a generic fetcher, an additional cachi2 package manager. This package manager utilizes a custom
+lockfile that is located in the input repository. Based on that lockfile, it will download files, save them into a requested
+location, and verify checksums. Below is a more detailed overview of the implementation.
+
+### Lockfile format
+
+Cachi2 expects the lockfile to be named `generic_lockfile.yaml`.
```

Comment:
It is possible, but I wonder what would be the benefit, if the lockfile has to be generated specifically for cachi2's use. But going off RPM package manager, it would probably be good to keep consistency and allow users to specify that name.

Reply:
One wouldn't need to check a random cachi2 only file to their upstream repo, but instead sideload it which makes IMO sense for formats which aren't ecosystem native or widely common.

Reply:
So, to clarify the format of the supplied path, should the input be an absolute path of the lockfile? I've looked over other package managers, and it seems like the pip package manager only allows relative paths, but that would defeat the purpose of sideloading the lockfile from outside the upstream repo. 

Just making sure if using an absolute path is okay.

Reply:
> but that would defeat the purpose of sideloading the lockfile from outside the upstream repo.

Yes, correct. That go me thinking though - we don't allow resources outside the repo in general, this would be a first IIRC. This would allow the operator to override a checked in version of the lockfile (because CLI should always take precedence) making us download something else. I don't think that's a problem though because everything would be recorded in the SBOM which might then not pass additional verification and we're not able to tell good from bad content pointed to by a URL today anyway, so from that POV there's no difference between a checked in version of the manifest vs an external one, so I think it's the right way forward.

Reply:
I still think this hurts our ability to enforce reproducibility. As you said, this is a behavior without precedents we'd be introducing. 

We could get into a debate about how much should Cachi2 be in the business of enforcing reproducibility, given that for some package managers (e.g. `go`), that claim is already somewhat flawed (because `go` can potentially resolve things differently based on many factors). I'd love to have a strong claim on reproducibility, in a way that we could cache a prefetch output based on repo + commit hash (and potentially Cachi2's version).

In any case, I think we could discuss this sideloading as a follow up, as it won't be a breaking change to the interface we're establishing here.

Reply:
> We could get into a debate about how much should Cachi2 be in the business of enforcing reproducibility, given that for some package managers (e.g. go), that claim is already somewhat flawed (because go can potentially resolve things differently based on many factors). I'd love to have a strong claim on reproducibility, in a way that we could cache a prefetch output based on repo + commit hash (and potentially Cachi2's version).

I think arguments passed via the CLI doesn't inherently doesn't contradict the claim, does it? Once you inject the same file, the outcome should not be different, otherwise it's a bug. If you inject a different one, then it's a different fetch/build, so we can't really talk about reproducibility, can we?

> In any case, I think we could discuss this sideloading as a follow up, as it won't be a breaking change to the interface we're establishing here.

I was wondering about this one. If we make it a follow up, then we inherently force (once we move this out of experimental before introducing the CLI option to pass an alternative manifest) early adopters to check in their manifest files to their upstream projects. I don't think they'll then be bothered much to revert that change since it'll work, you know, renovate/dependabot stuff would be set up... So I'm tempted to say that the CLI extra arg work should happen before the experimental status is lifted.

Reply:
> See [this PR](https://github.com/red-hat-data-services/trustyai-explainability/pull/44/files) that does sideloading that way

Thank you for pointing this out - I think for the sake of higher SLSA levels and build artifact confidence we should treat this as a bug and force a clean repo usage.

Reply:
> I was unable to find a clear definition for what a "dirty" repository is

Simple, git knows about untracked files - if not empty OR there are staged/unstaged changes -> dirty repo.

> but couldn't that be avoided by just committing the file locally?

I'm not sure I follow.

Reply:
> I'm not sure I follow.
Instead of just copying that file into the repo, once could also actually commit it locally in the task, without actually pushing it anywhere. 

Something like this:
```
cp /tmp/generic_lockfile.yaml .
git config --global user.name "Foo Bar"
git config --global user.email "foo.bar@example.com"
git add generic_lockfile.yaml
git commit -m "Add file"
```
This would leave a clean repo.

Reply:
Oh, not a cachi2 problem. If one is determined enough to go and use such a dirty hack in their pipeline they're should be ready to reap the potential consequences, there's nothing we can do about this in cachi2 and - on a side note - I'd argue noone should even try providing support for pipeline configurations like that.

Reply:
Updated to support specifying an absolute lockfile path as part of the JSON input.